opentech/apply/projects/views.py

```diff
@@ -11,20 +12,81 @@
 from opentech.apply.activity.messaging import MESSAGES, messenger
 from opentech.apply.activity.views import ActivityContextMixin, CommentFormView
 from opentech.apply.users.decorators import staff_required
-from opentech.apply.utils.views import (DelegateableView, DelegatedViewMixin,
-                                        ViewDispatcher)
+from opentech.apply.utils.views import (
+    DelegateableView,
+    DelegatedViewMixin,
+    ViewDispatcher
+)
 
 from .forms import (
+    ApproveContractForm,
     CreateApprovalForm,
     ProjectApprovalForm,
     ProjectEditForm,
     RejectionForm,
     RemoveDocumentForm,
     SetPendingForm,
     UpdateProjectLeadForm,
-    UploadDocumentForm,
+    UploadContractForm,
+    UploadDocumentForm
 )
-from .models import CONTRACTING, Approval, Project, PacketFile
+from .models import CONTRACTING, Approval, Contract, PacketFile, Project
+
+
+class ContractsMixin:
+    def get_context_data(self, **kwargs):
+        project = self.get_object()
+        recent_contracts = list(project.contracts.filter(is_signed=True)
+                                                 .order_by('-created_at')[:5])
+
+        # display an unsigned contract if it's more recent than the signed ones
+        latest_unsigned = project.contracts.filter(is_signed=False).first()
+        if latest_unsigned and latest_unsigned.created_at < recent_contracts[0]:
+            recent_contracts = [latest_unsigned, *recent_contracts[:4]]
+
+        try:
+            latest_contract = recent_contracts[0]
+        except IndexError:
+            latest_contract = None
+
+        context = super().get_context_data(**kwargs)
+        context['latest_contract'] = latest_contract
+        context['recent_contracts'] = recent_contracts
+        return context
+
+
+@method_decorator(staff_required, name='dispatch')
+class ApproveContractView(DelegatedViewMixin, UpdateView):
+    context_name = 'approve_contract_form'
+    form_class = ApproveContractForm
+    model = Contract
+    pk_url_kwarg = 'contract_pk'
+
+    def dispatch(self, request, *args, **kwargs):
+        try:
+            self.project = Project.objects.get(pk=self.kwargs['pk'])
```

Comment:
this has already been done with `kwargs['object']` i'm pretty sure?

aside: why not `get_object_or_404`?

Reply:
It has, but for some reason kwargs was missing the `object` key by the time this method was called.  I got as far as confirming the MRO was running this in the place I expected (after `DelegateableView.post`) but opted for the simpler fix.

aside: I originally had the except doing something else and didn't thing to rewrite, fixed and rebased.